### PR TITLE
Infer bounds for BatchNorm* Ops.

### LINKED ISSUE #809: Handle bounds in `BatchNormGradOp`shape function

### LINKED ISSUE #810: Handle bounds in `BatchNormInferenceOp`shape function

### LINKED ISSUE #811: Handle bounds in `BatchNormTrainingOp`shape function

### LINKED ISSUE #869: Migrate away from AllShapesMatch/AllTypesMatch in BatchNorm ops
BatchNorm ops' ODS currently says something along the lines of:

```td
def StableHLO_BatchNormInferenceOp : StableHLO_Op<"batch_norm_inference",
    [Pure, AllTypesMatch<["operand", "result"]>,
    AllShapesMatch<["scale", "offset", "mean", "variance"]>,
    InferTensorType]> {
  ...
}
```

This was working fine before we introduced the notion of `isCompatibleForHloTypeInference`, but nowadays this is too restrictive. We should review all usages of `AllShapesMatch` and `AllTypesMatch` in BatchNorm ops and migrate away from them to make sure that we only verify shape/type compatibility, not equality.

### PR DESCRIPTION
**Pull request by subhankarshah**
closes #809, closes #810, closes #811, closes #869

Unify verification and inference for BatchNorm* Ops

  * For all `batch_norm_*` ops infer all shapes and bounds from `operand` only.
  * Verify `operand` feature dimension compatibility with `dim[0]` of `scale`.
  * Remove `AllShapesMatch` and convert `AllTypesMatch` to `AllElementTypesMatch`.
  * Check for shape compatibility instead of equality.

### CODE REVIEW

File: stablehlo/tests/infer_stablehlo.mlir [hunk 1]
```diff
@@ -262,6 +262,21 @@ func.func @batch_norm_inference(%input: tensor<4x256xf32>, %scale: tensor<256xf3
 
 // -----
 
+// CHECK-LABEL: @batch_norm_inference_dynamic
+func.func @batch_norm_inference_dynamic(
```
Comment (zhouxin913) [inline stablehlo/tests/infer_stablehlo.mlir:266]: Could you add "_bounds" in the method name, like all other bounds tests? I admit the test file definitely need some cleaning up but before that let's keep more things aligned.

File: stablehlo/tests/infer_stablehlo.mlir [hunk 2]
```diff
@@ -262,6 +262,21 @@ func.func @batch_norm_inference(%input: tensor<4x256xf32>, %scale: tensor<256xf3
...
+  %input: tensor<4x?xf32, #stablehlo.type_extensions<bounds = [?, 64]>>, %scale: tensor<?xf32>,
+  %offset: tensor<?xf32>, %mean: tensor<?xf32>, %variance: tensor<?xf32>
+) -> (tensor<*xindex>) {
+  %0 = "stablehlo.batch_norm_inference" (%input, %scale, %offset, %mean, %variance) {
+    epsilon = 1.001000e-05 : f32, feature_index = 1 : i64
+    } : (tensor<4x?xf32, #stablehlo.type_extensions<bounds = [?, 64]>>, tensor<?xf32>, tensor<?xf32>, tensor<?xf32>, tensor<?xf32>) -> tensor<4x?xf32, #stablehlo.type_extensions<bounds = [?, 64]>>
+  // CHECK: (tensor<4x?xf32, #stablehlo.type_extensions<bounds = [?, 64]>>) -> tensor<*xindex>
```
Comment (zhouxin913) [inline stablehlo/tests/infer_stablehlo.mlir:273]: Let's write it like this?
// CHECK: types0 = tensor<...

**Review by zhouxin913 — CHANGES_REQUESTED** [review]

File: stablehlo/tests/infer_stablehlo.mlir (hunk 2 above)
Comment (subhankarshah) [reply stablehlo/tests/infer_stablehlo.mlir:273]: That is not possible for this op.

Comment (subhankarshah) [reply stablehlo/tests/infer_stablehlo.mlir:273]: Get return types was not working as expected for this op, which is why I had to do it this way.

Comment (subhankarshah) [reply stablehlo/tests/infer_stablehlo.mlir:273]: Also the reason why I raised #902 .

**Comment by subhankarshah** [comment]
@smit-hinsu PTAL.

File: stablehlo/dialect/TypeInference.cpp [hunk 3]
```diff
@@ -1344,47 +1344,59 @@ LogicalResult inferAllToAllOp(
...
-LogicalResult inferBatchNormInferenceOp(
+LogicalResult inferBatchNorm(
     Optional<Location> location, Value operand, Value scale,
     int64_t featureIndex,
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  if (failed(verifyBatchNorm(location, operand, scale, featureIndex)))
+  if (failed(inferBatchNormOperand(location, operand, scale, featureIndex,
+                                   inferredReturnShapes)))
     return failure();
   auto operandType = operand.getType().cast<RankedTensorType>();
```
Comment (smit-hinsu) [inline stablehlo/dialect/TypeInference.cpp:1367]: add a comment that op definition requires ranked operands.

**Review by smit-hinsu — COMMENTED** [review]
Thanks!

This PR removes some duplication which is good but now is a bit unintuitive and requires a closer look. Like, why only inferBatchNormOperand does the verification, what does operand mean in the name, etc.

How about just having a single infer method that takes an extra parameter is_inference_op? Downside is that this is inconsistent with the president of one infer for each op. We can avoid that by calling this single helper from all three.

File: stablehlo/dialect/TypeInference.cpp [hunk 4]
```diff
@@ -1344,47 +1344,60 @@ LogicalResult inferAllToAllOp(
   return success();
 }
 
-LogicalResult inferBatchNormGradOp(
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1347]: Agreed with Smit about deduplicating into a single helper function with `is_inference`. We also have a separate section of the file for helpers (above line 1260) which would be a good place for this function.

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/dialect/TypeInference.cpp [hunk 5]
```diff
@@ -1344,46 +1361,39 @@ LogicalResult inferAllToAllOp(
...
 
-LogicalResult inferBatchNormTrainingOp(
-    Optional<Location> location, Value operand, Value scale,
-    int64_t featureIndex,
-    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  if (failed(verifyBatchNorm(location, operand, scale, featureIndex)))
-    return failure();
-  auto operandType = operand.getType().cast<RankedTensorType>();
-  inferredReturnShapes.emplace_back(operandType.cast<ShapedType>());
+  inferredReturnShapes.emplace_back(
```
Comment (smit-hinsu) [inline stablehlo/dialect/TypeInference.cpp:1390]: nit: we could construct once and push back twice so it is clear both have the same type.

**Review by smit-hinsu — APPROVED** [review]
Shape functions look good to me. Haven't looked into the verification change.

**Comment by subhankarshah** [comment]
@burmako Should we also update the spec for BatchNorm* Ops to say compatible shapes instead of same shapes, also define what compatible shapes mean.

File: stablehlo/dialect/TypeInference.cpp [hunk 6]
```diff
@@ -1344,46 +1361,37 @@ LogicalResult inferAllToAllOp(
...
-  SmallVector<int64_t> featureShape{featureCount};
-  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
-  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
-  return success();
-}
+  // Batch norm ops require operands to be ranked.
+  auto operandType = multiDimOperands[0].getType().cast<RankedTensorType>();
+  inferredReturnShapes.emplace_back(
+      operandType.getShape(), operandType.getElementType(),
+      operandType.cast<RankedTensorType>().getEncoding());
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1377]: This cast is not needed here.

File: stablehlo/dialect/TypeInference.cpp [hunk 7]
```diff
@@ -1344,46 +1361,37 @@ LogicalResult inferAllToAllOp(
...
-  inferredReturnShapes.emplace_back(operandType.cast<ShapedType>());
 
-  const int64_t featureCount = operandType.getDimSize(featureIndex);
-  SmallVector<int64_t> featureShape{featureCount};
-  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
-  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
-  return success();
-}
+  // Batch norm ops require operands to be ranked.
+  auto operandType = multiDimOperands[0].getType().cast<RankedTensorType>();
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1374]: `multiDimType`, since there are both multi-dim and single-dim operands in scope?

File: stablehlo/dialect/TypeInference.cpp [hunk 8]
```diff
@@ -1344,46 +1361,37 @@ LogicalResult inferAllToAllOp(
...
 
-LogicalResult inferBatchNormTrainingOp(
-    Optional<Location> location, Value operand, Value scale,
-    int64_t featureIndex,
-    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  if (failed(verifyBatchNorm(location, operand, scale, featureIndex)))
-    return failure();
-  auto operandType = operand.getType().cast<RankedTensorType>();
-  inferredReturnShapes.emplace_back(operandType.cast<ShapedType>());
+  SmallVector<int64_t> featureShape{operandType.getDimSize(featureIndex)};
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1381]: `singleDimShape`, to be consistent with the multi-dim and single-dim nomenclature? (Same for `operandBounds` => `multiDimBounds`, `featureBounds` => `singleDimBounds`, `singleDimensionResultShape` => `singleDimShape`).

File: stablehlo/dialect/TypeInference.cpp [hunk 9]
```diff
@@ -1344,46 +1361,37 @@ LogicalResult inferAllToAllOp(
...
+
+  SmallVector<int64_t> operandBounds =
+      to_vector(encodingToBounds(operandType.getEncoding()));
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1384]: Why do we need `to_vector` here?

File: stablehlo/dialect/TypeInference.cpp [hunk 10]
```diff
@@ -1344,46 +1361,37 @@ LogicalResult inferAllToAllOp(
...
+  SmallVector<int64_t> featureBounds;
+  if (!operandBounds.empty())
+    featureBounds.emplace_back(operandBounds[featureIndex]);
+
+  auto singleDimensionResultShape = ShapedTypeComponents(
+      featureShape, operandType.getElementType(),
+      boundsToEncoding(operandType.getEncoding(), featureBounds));
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1391]: Does this mean that the result type will inherit `operandType.getEncoding()` if bounds are empty? This doesn't sound obviously correct. `featureBounds.empty() ? nullptr : boundsToEncoding(...)` looks more conservative.

File: stablehlo/dialect/TypeInference.cpp [hunk 11]
```diff
@@ -199,9 +199,28 @@ int64_t stridedBound(int64_t bound, int64_t windowSize, int64_t stride) {
...
+    return emitOptionalError(
+        location,
+        "expects multi-dimensional operands to have compatible shapes.");
+
+  if (failed(verifyPairwiseCompatibleShapes(singleDimOperands.getTypes())))
+    return emitOptionalError(
+        location,
+        "expects single-dimensional operands to have compatible shapes.");
+
+  auto operandType = multiDimOperands[0].getType().cast<RankedTensorType>();
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:223]: Same suggestion regarding renaming as in the functions below. Given that we have the new nomenclature that involves single-dim and multi-dim operands, let's change all the old names to follow it. Otherwise, stuff like `operandType` and `scaleShape` looks out of place.

**Review by burmako — CHANGES_REQUESTED** [review]
Nice idea to deduplicate things! I left some comments on the details of the underlying refactoring.

File: stablehlo/dialect/TypeInference.cpp [hunk 12]
```diff
@@ -212,16 +231,14 @@ LogicalResult verifyBatchNorm(Optional<Location> location, Value operand,
...
-      scale.getType().cast<RankedTensorType>().getDimSize(0);
-  // As ODS enforces `scale`, `mean`, `variance`, `offset` are AllShapesMatch,
-  // this also infers that featureCount is aligned with them.
-  if (scaleShape != featureCount)
+      singleDimOperands[0].getType().cast<RankedTensorType>().getDimSize(0);
+
+  if (!verifyCompatibleDims(scaleShape, featureCount))
     return emitOptionalError(
         location,
         "expects the size of scale factor to be same as the "
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:244]: "to be compatible with"

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/tests/ops_stablehlo.mlir [hunk 13]
```diff
@@ -4847,22 +4847,14 @@ func.func @error_batch_norm_grad(%input: tensor<2x2x2x2xf32>, %scale: tensor<2xf
...
 // -----
 
-func.func @error_batch_norm_grad(%input: tensor<*xf32>, %scale: tensor<2xf32>, %mean: tensor<2xf32>, %variance: tensor<2xf32>, %grad_output: tensor<2x2x2x2xf32>) -> tensor<2x2x2x2xf32> {
+func.func @error_batch_norm_grad(%input: tensor<*xf32>, %scale: tensor<2xf32>, %mean: tensor<2xf32>, %variance: tensor<2xf32>, %grad_output: tensor<2x2x2x2xf32>) -> tensor<*xf32> {
   // expected-error@+1 {{operand #0 must be ranked tensor of f8E4M3FN type or f8E5M2 type or 16-bit float or 32-bit float or 64-bit float or bfloat16 type values, but got 'tensor<*xf32>'}}
   %0:3 = "stablehlo.batch_norm_grad" (%input, %scale, %mean, %variance, %grad_output) {epsilon = 0.001 : f32, feature_index = 0 : i64} : (tensor<*xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2x2x2x2xf32>) -> (tensor<*xf32>, tensor<2xf32>, tensor<2xf32>)
   func.return %0#0 : tensor<*xf32>
 }
 
 // -----
```
Comment (ghpvnist) [inline stablehlo/tests/ops_stablehlo.mlir:4856]: Let's also add theses tests for dynamic shapes:
```
// CHECK-LABEL: @batch_norm_train_dynamic
func.func @batch_norm_train_dynamic(%input: tensor<?x?x2x2xf32>, %scale: tensor<2xf32>, %offset: tensor<2xf32>) -> tensor<?x?x2x2xf32> {
  %0:3 = "stablehlo.batch_norm_training" (%input, %scale, %offset) {epsilon = 0.001 : f32, feature_index = 1 : i64} : (tensor<?x?x2x2xf32>, tensor<2xf32>, tensor<2xf32>) -> (tensor<?x?x2x2xf32>, tensor<2xf32>, tensor<2xf32>)
  func.return %0#0 : tensor<?x?x2x2xf32>
}

// -----

// CHECK-LABEL: @batch_norm_inference_dynamic
func.func @batch_norm_inference_dynamic(%input: tensor<4x?xf32>, %scale: tensor<?xf32>, %offset: tensor<?xf32>, %mean: tensor<?xf32>, %variance: tensor<?xf32>) -> (tensor<4x?xf32>) {
  %0 = "stablehlo.batch_norm_inference" (%input, %scale, %offset, %mean, %variance) {epsilon = 1.001000e-05 : f32, feature_index = 1 : i64} :
      (tensor<4x?xf32>, tensor<?xf32>, tensor<?xf32>, tensor<?xf32>,
        tensor<?xf32>) -> tensor<4x?xf32>
  func.return %0 : tensor<4x?xf32>
}

// -----

func.func @batch_norm_grad_dynamic(%input: tensor<?x2x2x2xf32>, %scale: tensor<?xf32>, %mean: tensor<?xf32>, %variance: tensor<?xf32>, %grad_output: tensor<?x2x2x2xf32>) -> tensor<?x2x2x2xf32> {
  %0:3 = "stablehlo.batch_norm_grad" (%input, %scale, %mean, %variance, %grad_output) {epsilon = 0.001 : f32, feature_index = 0 : i64} : (tensor<?x2x2x2xf32>, tensor<?xf32>, tensor<?xf32>, tensor<?xf32>, tensor<?x2x2x2xf32>) -> (tensor<?x2x2x2xf32>, tensor<?xf32>, tensor<?xf32>)
  func.return %0#0 : tensor<?x2x2x2xf32>
}
```

**Review by ghpvnist — CHANGES_REQUESTED** [review]

**Review by ghpvnist — APPROVED** [review]

File: stablehlo/dialect/TypeInference.h [hunk 14]
```diff
@@ -112,18 +112,18 @@ LogicalResult inferAllToAllOp(
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
 
 LogicalResult inferBatchNormGradOp(
-    Optional<Location> location, Value operand, Value scale,
-    int64_t featureIndex,
+    Optional<Location> location, Value operand, Value scale, Value mean,
+    Value variance, Value gradOutput, int64_t feature_index,
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.h:116]: `featureIndex` (Here and everywhere else).

File: stablehlo/dialect/TypeInference.cpp [hunk 15]
```diff
@@ -199,39 +199,92 @@ int64_t stridedBound(int64_t bound, int64_t windowSize, int64_t stride) {
...
+        "expects single-dimensional operands to have compatible shapes.");
+
+  auto multiDimType = multiDimOperands[0].getType().cast<RankedTensorType>();
+  if (feature_index >= multiDimType.getRank())
     return emitOptionalError(
         location,
         "expects feature_index to be smaller than the rank of "
-        "operand type; got feature_index ",
-        feature_index, ", and rank ", operandType.getRank(), ".");
+        "multi-dimensional operand; got feature_index ",
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:228]: Nit: "multi-dimensional operands".

File: stablehlo/dialect/TypeInference.cpp [hunk 16]
```diff
@@ -199,39 +199,92 @@ int64_t stridedBound(int64_t bound, int64_t windowSize, int64_t stride) {
...
-  // As ODS enforces `scale`, `mean`, `variance`, `offset` are AllShapesMatch,
-  // this also infers that featureCount is aligned with them.
-  if (scaleShape != featureCount)
+  if (!verifyCompatibleDims(singleDimSize, featureCount))
     return emitOptionalError(
         location,
-        "expects the size of scale factor to be same as the "
-        "feature count, but the size of scale factor is ",
-        dimSizeToString(scaleShape), " and the feature count is ",
+        "expects the size of single-dimensional operand to be compatible with "
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:244]: Nit: "single-dimensional operands".

**Review by burmako — APPROVED** [review]

**Review by burmako — APPROVED** [review]
